pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -1052,6 +1059,52 @@ func pruneBundle(tx *gorm.DB, trustDomainID string, expiry time.Time, log logrus
 	return changed, nil
 }
 
+func revokeX509CA(tx *gorm.DB, trustDomainID string, publicKey crypto.PublicKey) error {
+	model := &Bundle{}
+	if err := tx.Find(model, "trust_domain = ?", trustDomainID).Error; err != nil {
+		return sqlError.Wrap(err)
+	}
+
+	bundle, err := modelToBundle(model)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to unmarshal bundle: %v", err)
+	}
+
+	found := false
+	var rootCAs []*common.Certificate
+	for _, ca := range bundle.RootCas {
+		cert, err := x509.ParseCertificate(ca.DerBytes)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to parse root CA: %v", err)
+		}
+
+		ok, err := cryptoutil.PublicKeyEqual(cert.PublicKey, publicKey)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to compare public key: %v", err)
+		}
+
+		if ok {
+			if !ca.TaintedKey {
+				return status.Error(codes.Internal, "it is not possible to revoke an untainted root CA")
```

Comment:
The error is due to an invalid input argument, not an internal failure.

```suggestion
				return status.Error(codes. InvalidArgument, "it is not possible to revoke an untainted root CA")
```